### PR TITLE
Major player rewrite + event service implementation

### PR DESCRIPTION
**Pull request by devgianlu**
This PR wants to implement the `event-service` framework because of https://github.com/librespot-org/librespot/issues/342, in order to accomodate the required metrics **a major player rewrite has also been done**.

All requests are sent to `hm://event-service/v1/events` via a `POST` request. The request has three headers: `X-ClientTimeStamp`, `Accept-Language`, `X-Offset`. I'm yet to figure out what the last one does. The body of the request is generally as follows: some chars to identify the type of message and the actual data. The `0x09` character is used to separate various fields.

I have also enhanced the Wireshark scripts to parse event service requests: https://github.com/devgianlu/spotify-analyze

List of client events: https://gist.github.com/devgianlu/8371de4449e90fc18e9bef1a0745bd18

TODO:
- [x] Re-implement crossfade
- [x] Fix repeat track not working
- [ ] Fix reporting when source end is different from `trackdone`
- [x] Implement known metrics

### CODE REVIEW

**Comment by backspac** [comment]
Just a wild guess. Following this doc, `X-Offset` might be the track position in an album or a playlist.
https://developer.spotify.com/documentation/web-api/reference/player/start-a-users-playback/

**Comment by devgianlu** [comment]
> Just a wild guess. Following this doc, `X-Offset` might be the track position in an album or a playlist.
> https://developer.spotify.com/documentation/web-api/reference/player/start-a-users-playback/

Doesn't look like the values are usually even higher than the playlist size.

**Comment by devgianlu** [comment]
**This is finally working!!** Except when the "ended how" field is different from `trackdone`. There is probably some more research to do to figure out what those extra fields are for.

**Comment by Yuhanun** [comment]
Hello,

Is there currently any way to report back to spotify that you listened to a sing using the API that this daemon provides?

Thank you,
Yuhanun

**Comment by devgianlu** [comment]
> Is there currently any way to report back to spotify that you listened to a song using the API that this daemon provides?

Yes, when this is going to be merged the client will have complete reporting functionalities.

**Comment by Yuhanun** [comment]
> > Is there currently any way to report back to spotify that you listened to a song using the API that this daemon provides?
> 
> Yes, when this is going to be merged the client will have complete reporting functionalities.

Any ETA? Is there any way to do this with the current branch?

**Comment by devgianlu** [comment]
@Yuhanun No idea when this will be finished. The protocol right know is under development.

**Comment by Yuhanun** [comment]
@devgianlu So right now there is absolutely *no* way to do the listen reports?

**Comment by devgianlu** [comment]
There is, but why would you need it outside of the player?

**Comment by Yuhanun** [comment]
Because I'm building my own CLI tool and I'd like to contribute to the artist their stream count.

File: core/src/main/java/xyz/gianlu/librespot/player/Player.java [hunk 1]
```diff
@@ -497,6 +477,30 @@ private void loadAutoplay() {
...
+    private void endMetrics(String playbackId, @NotNull PlaybackMetrics.Reason reason, @Nullable PlayerMetrics playerMetrics, int when) {
+        if (playbackId == null) return;
+
+        PlaybackMetrics pm = metrics.remove(playbackId);
+        if (pm == null) return;
+
+        pm.endedHow(reason, state.getPlayOrigin().getFeatureIdentifier());
+        pm.endInterval(when);
+        pm.update(playerMetrics);
+        session.eventService().trackPlayed(pm);
```
Comment (Yuhanun) [inline core/src/main/java/xyz/gianlu/librespot/player/Player.java:500]: trackPlayed called in endMetrics

File: core/src/main/java/xyz/gianlu/librespot/player/Player.java [hunk 2]
```diff
@@ -554,20 +558,12 @@ public void trackChanged(@NotNull String playbackId, @Nullable TrackOrEpisode me
...
-    public void trackPlayed(@NotNull PlaybackMetrics.Reason endReason, @NotNull PlayerMetrics playerMetrics, int willEndAt) {
-        if (metrics != null) {
-            metrics.endedHow(endReason, state.getPlayOrigin().getFeatureIdentifier());
-            metrics.endInterval(willEndAt);
-            metrics.update(playerMetrics);
-            session.eventService().trackPlayed(metrics);
-            metrics = null;
-        }
+    public void trackPlayed(@NotNull String playbackId, @NotNull PlaybackMetrics.Reason endReason, @NotNull PlayerMetrics playerMetrics, int when) {
+        endMetrics(playbackId, endReason, playerMetrics, when);
```
Comment (Yuhanun) [inline core/src/main/java/xyz/gianlu/librespot/player/Player.java:566]: endMetrics called in trackPlayed

Comment (devgianlu) [reply core/src/main/java/xyz/gianlu/librespot/player/Player.java:566]: That's not the same method, they just have the same name.

Comment (Yuhanun) [reply core/src/main/java/xyz/gianlu/librespot/player/Player.java:566]: yeah my bad, sorry :)

**Comment by Yuhanun** [comment]
Heyo this doesn't work for me, it doesn't report back "audio streams" (like the listens counters) to spotify, it never gest updated on their end even when listening for 60 seconds+

**Comment by hbouhadji** [comment]
+1 @Yuhanun

**Comment by Shinoby92** [comment]
+1 still not fixed? anywhere to figure out and help to retrieve how the original workflow works?

**Comment by devgianlu** [comment]
> +1 still not fixed? anywhere to figure out and help to retrieve how the original workflow works?

It kind of works, but it's not perfect. If you want to try and analyze the request chain you need to hookup spotify-analyze and dump the Mercury requests.

**Pull request merged**